src/simple_uds.rs

```diff
@@ -0,0 +1,185 @@
+        match e {
+            Error::Json(e) => ::Error::Json(e),
+            e => ::Error::Transport(Box::new(e)),
+        }
+    }
+}
+
+impl ::std::error::Error for Error {}
+
+/// Simple synchronous UDS transport.
+#[derive(Debug, Clone)]
+pub struct UdsTransport {
+    /// The path to the Unix Domain Socket
+    pub sockpath: path::PathBuf,
+    /// The read and write timeout to use
+    pub timeout: Option<time::Duration>,
+}
+
+impl UdsTransport {
+    /// Create a new UdsTransport without timeouts to use
+    pub fn new<P: AsRef<path::Path>>(sockpath: P) -> UdsTransport {
+        UdsTransport {
+            sockpath: sockpath.as_ref().to_path_buf(),
+            timeout: None,
+        }
+    }
+
+    fn request<R>(&self, req: impl serde::Serialize) -> Result<R, Error>
+    where
+        R: for<'a> serde::de::Deserialize<'a>,
+    {
+        let mut sock = UnixStream::connect(&self.sockpath)?;
+        sock.set_read_timeout(self.timeout)?;
+        sock.set_write_timeout(self.timeout)?;
+
+        serde_json::to_writer(&mut sock, &req)?;
+
+        // NOTE: we don't check the id there, so it *must* be synchronous
+        let resp: R = serde_json::Deserializer::from_reader(&mut sock)
+            .into_iter()
+            .next()
+            .ok_or(Error::Timeout)??;
+        Ok(resp)
+    }
+}
+
+impl Transport for UdsTransport {
+    fn send_request(&self, req: Request) -> Result<Response, ::Error> {
+        Ok(self.request(req)?)
+    }
+
+    fn send_batch(&self, reqs: &[Request]) -> Result<Vec<Response>, ::Error> {
+        Ok(self.request(reqs)?)
+    }
+
+    fn fmt_target(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.sockpath.to_string_lossy())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    #[cfg(not(windows))]
+    use std::os::unix::net::UnixListener;
+    #[cfg(windows)]
+    use uds_windows::UnixListener;
+
+    use std::{
+        fs,
+        io::{Read, Write},
+        thread,
+    };
+
+    use super::*;
+    use Client;
+
+    // Test a dummy request / response over an UDS
+    #[test]
+    fn sanity_check_uds_transport() {
+        let socket_path: path::PathBuf = "uds_scratch.socket".into();
+        // Any leftover?
+        fs::remove_file(&socket_path).unwrap_or_else(|_| ());
+
+        let server = UnixListener::bind(&socket_path).unwrap();
+        let dummy_req = Request {
+            method: "getinfo",
+            params: &[],
+            id: serde_json::Value::Number(111.into()),
+            jsonrpc: Some("2.0".into()),
+        };
+        let dummy_req_ser = serde_json::to_vec(&dummy_req).unwrap();
+        let dummy_resp = Response {
+            result: None,
+            error: None,
+            id: serde_json::Value::Number(111.into()),
+            jsonrpc: Some("2.0".into()),
+        };
+        let dummy_resp_ser = serde_json::to_vec(&dummy_resp).unwrap();
+
+        let cli_socket_path = socket_path.clone();
+        let client_thread = thread::spawn(move || {
```

Comment:
This code is correct (it seems) when the test works OK, but it can act meh in case it actually find some non-trivial bug somewhere. In the best case - it will just leave a detached thread running, which is OKis, I guess, but in the worse, it might hang on the `.join` later, making the user wonder what have happened, and why is not finishing.

The simplest solution is to just bound the loop to some reasonable value. Since the retry delay is fixed, assuming that this thread shouldn't ever need more than 30s will at least guarantee that the test terminates.

The best structure is using `Arc<AtomicBool>` or something like that to signal to the thread loop termination request, and wrapping the join handle into some newtype, that on drop: sets the termination flag and `join` s on the handle. This way the no bug can hang the thread, and no bug can miss joining on it.

Reply:
Ah i was going to comment "hehe, it's not" but i actually forgot to set the timeout in the constructor of the transport. With it you wouldn't block for more than *timeout* seconds on `join`.

Reply:
Added a timeout in the Transport as mentioned above, and removed the loop. Tested by commenting out the response write from the main thread:
```diff
diff --git a/src/simple_tcp.rs b/src/simple_tcp.rs
index 7a6dfdb..1018f14 100644
--- a/src/simple_tcp.rs
+++ b/src/simple_tcp.rs
@@ -137,19 +137,13 @@ mod tests {
         let dummy_resp_ser = serde_json::to_vec(&dummy_resp).unwrap();
 
         let client_thread = thread::spawn(move || {
-            let transport = TcpTransport::new(addr);
+            let transport = TcpTransport {
+                addr,
+                timeout: Some(time::Duration::from_secs(5)),
+            };
             let client = Client::with_transport(transport);
 
-            loop {
-                match client.send_request(dummy_req.clone()) {
-                    Ok(resp) => return resp,
-                    Err(e) => {
-                        // Print so should it fail it can be debugged with --nocapture
-                        eprintln!("Error while sending request: '{:?}'", e);
-                        thread::sleep(time::Duration::from_millis(100));
-                    }
-                }
-            }
+            client.send_request(dummy_req.clone()).unwrap()
         });
 
         let (mut stream, _) = server.accept().unwrap();
@@ -161,8 +155,8 @@ mod tests {
         }
         assert_eq!(recv_req, dummy_req_ser);
 
-        stream.write(&dummy_resp_ser).unwrap();
-        stream.flush().unwrap();
+        //stream.write(&dummy_resp_ser).unwrap();
+        //stream.flush().unwrap();
         let recv_resp = client_thread.join().unwrap();
         assert_eq!(serde_json::to_vec(&recv_resp).unwrap(), dummy_resp_ser);
     }
diff --git a/src/simple_uds.rs b/src/simple_uds.rs
index 9e506f1..69881bd 100644
--- a/src/simple_uds.rs
+++ b/src/simple_uds.rs
@@ -149,19 +149,13 @@ mod tests {
 
         let cli_socket_path = socket_path.clone();
         let client_thread = thread::spawn(move || {
-            let transport = UdsTransport::new(cli_socket_path);
+            let transport = UdsTransport {
+                sockpath: cli_socket_path,
+                timeout: Some(time::Duration::from_secs(5)),
+            };
             let client = Client::with_transport(transport);
 
-            loop {
-                match client.send_request(dummy_req.clone()) {
-                    Ok(resp) => return resp,
-                    Err(e) => {
-                        // Print so should it fail it can be debugged with --nocapture
-                        eprintln!("Error while sending request: '{:?}'", e);
-                        thread::sleep(time::Duration::from_millis(100));
-                    }
-                }
-            }
+            client.send_request(dummy_req.clone()).unwrap()
         });
 
         let (mut stream, _) = server.accept().unwrap();
```